Zephyr: Add ADC support.

bikeNomad opened the pull request:
### Summary

This PR adds support for ADC peripherals in the Zephyr port.

As is typical for Zephyr, the ADC channel setup is done in the devicetree (typically using an overlay).

This code requires ADC channels to be listed in the `io-channels` property of the `zephyr,user` root node like this:
```
/ {
	zephyr,user {
		io-channels = <&adc 0>, <&adc 1>, <&adc 4>, <&adc 5>, <&adc 7>;
	};
};

&adc {
	#address-cells = <1>;
	#size-cells = <0>;

	channel@0 {
		reg = <0>;
		zephyr,gain = "ADC_GAIN_1_6";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,input-positive = <NRF_SAADC_AIN0>; /* P0.02 */
		zephyr,resolution = <12>;
	};
       // other channels defined here (1, 4, 5, 7)
};
```


### Testing

I have tested this on the Seeed Studio XIAO BLE NRF52840 SENSE board.

### Trade-offs and Alternatives

This code may be enabled and disabled by the usual means in `mpconfigport.h`:
```
#define MICROPY_HW_ENABLE_ADC (1)
#define MICROPY_PY_MACHINE_ADC (1)
#define MICROPY_PY_MACHINE_ADC_INCLUDEFILE "ports/zephyr/machine_adc.c"
#define MICROPY_PY_MACHINE_ADC_READ (1)
#define MICROPY_PY_MACHINE_ADC_READ_UV (1)
```

bikeNomad commented:
This PR includes @dpgeorge 's PRs #17610, #17599, 17552. I don't think it'll work as well without those.

bikeNomad commented:
> I've now merged those other PRs, so please can you rebase this PR on latest master?

Done.

dpgeorge commented on ports/zephyr/README.md:
```diff
@@ -153,6 +154,13 @@ Example of using SPI to write a buffer to the MOSI pin:
+
+    from machine import ADC
+
+    adc = ADC(('adc', 0))
```
Please use double quotes `ADC(("adc", 0))` to match examples above.

michalpiekos reviewed (approved):
LGTM

michalpiekos commented on ports/zephyr/machine_adc.c:
```diff
@@ -0,0 +1,238 @@
+#endif
+
+#define DT_SPEC_AND_COMMA(node_id, prop, idx) \
+    ADC_DT_SPEC_GET_BY_IDX(node_id, idx),
```
I am working on DAC extension for zephyr port. It also uses io-channels. Your code would not work if some of the io-channels are actually DAC channels.

ned-pcs replied:
What do you recommend?

michalpiekos replied:
Below is what I did for DAC. It should be easy to re-work to ADC. `make_new` takes index of a channel or channel name then I check if `dac_channels` has proper channel under given index or INVALID_CHANNEL. This is how I know it is DAC.
If DAC object is initialized with channel name then I check its index in `io_channel_names` and the rest is as above.

If something is unclear or you find some improvement just let me know. It is still work in progress.

#define INVALID_CHANNEL (-1)                                                                                                                                                                                           
#define N_CHANNELS  DT_PROP_LEN(USER_NODE, io_channels)                                                                                                                                                                
#define NAME_ELEM(node, prop, idx)  [idx] = DT_PROP_BY_IDX(node, prop, idx),                                                                                                                                           
#define IO_DEV(node, prop, idx)  DEVICE_DT_GET(DT_PHANDLE_BY_IDX(node, prop, idx)),
#define IS_DAC(node, prop, idx)  DT_PHA_HAS_CELL_AT_IDX(node, prop, idx, output),
#define DAC_ONLY(node_id, prop, idx)                                         \
    DT_PHA_BY_IDX_OR(node_id, prop, idx, output, INVALID_CHANNEL),

static const struct device *io_devs[N_CHANNELS] = {
    DT_FOREACH_PROP_ELEM(USER_NODE, io_channels, IO_DEV)
};

static const char *io_channel_names[] = {
    DT_FOREACH_PROP_ELEM(USER_NODE, io_channel_names, NAME_ELEM)
};

static const bool io_is_dac[N_CHANNELS] = {
    DT_FOREACH_PROP_ELEM(USER_NODE, io_channels, IS_DAC)
};

static const int dac_channels[] = {
    DT_FOREACH_PROP_ELEM(USER_NODE, io_channels, DAC_ONLY)
};

bikeNomad replied:
Why not have different properties in the `/zephyr,user` node? Is there any advantage to using `io-channels` to refer to both the ADC and DAC channels? Couldn't we do something like this:
```
 / {
 	zephyr,user {
 		adc-channels = <&adc 0>, <&adc 1>, <&adc 4>, <&adc 5>, <&adc 7>;
 		dac-channels = <&dac 0>, <&dac 1>;
 	};
 };
```

ned-pcs replied:
@michalpiekos How does your `IS_DAC()` macro work? What would be an equivalent `IS_ADC()` macro?

bikeNomad commented:
> The problem is that `frdm_k64f` doesn't have `io-channels` defined.
> 
> But it does have an ADC and does define `adc0_default` and `adc1_default` in its device overlay/tree. So do many other boards.
> 
> Can you make it so that the ADC driver here uses `adc0_default` (and `adc1_default` etc) so it just works out of the box with a majority of the Zephyr boards?

The issue is a little more complicated than that, as @michalpiekos points out. I can easily make it so that the stock `frdm_k64f` compiles OK by checking for the existence of `io-channels` within the `/zephyr,user` node.  This would be something that would be set for a specific build (so that MicroPython, for instance, could have a `frdm_k64f.overlay` that provides access to specific ADC channels in the `io-channels` list. But there's some magic in that label, because `#io-channel-cells` is defined within the ADC and DAC nodes.

`adc0_default` and `adc1_default` refer to something else (the ADCs in the FRDM_K64F are named `adc0` and `adc1`).

I see some ADC channels being referred to in the same DTS from FRDM_K64F, but I _think_ we'd see these in the `zsensor` interface:
```
	temp0: temp0 {
		compatible = "nxp,kinetis-temperature";
		io-channels = < &adc0 0x1a >, < &adc0 0x1b >;
		io-channel-names = "SENSOR", "BANDGAP";
		bandgap-voltage = < 0xf4240 >;
		vtemp25 = < 0xaece0 >;
		sensor-slope-cold = < 0x654 >;
		sensor-slope-hot = < 0x654 >;
		status = "disabled";
	};
	temp1: temp1 {
		compatible = "nxp,kinetis-temperature";
		io-channels = < &adc1 0x1a >, < &adc1 0x1b >;
		io-channel-names = "SENSOR", "BANDGAP";
		bandgap-voltage = < 0xf4240 >;
		vtemp25 = < 0xaece0 >;
		sensor-slope-cold = < 0x654 >;
		sensor-slope-hot = < 0x654 >;
		status = "okay";
	};
```

I will try to figure out @michalpiekos 's code and adapt it for the ADC so we can both use the `io-channels` list.

bikeNomad committed:
zephyr/machine_adc: Add ADC support.

This commit adds support for ADC peripherals in the Zephyr port.

As is typical for Zephyr, the ADC channel setup is done in the devicetree
(typically using an overlay).  This code requires ADC channels to be listed
in the io-channels property of the zephyr,user root node.

Signed-off-by: Ned Konz <ned@metamagix.tech>

dpgeorge reviewed (approved):
Thanks for updating based on the review feedback.
